### PR TITLE
feat!: drop support for function-style rules and rules missing schemas

### PR DESCRIPTION
**Pull request by bmish**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[x] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)

This implements [RFC-85](https://github.com/eslint/rfcs/tree/main/designs/2021-schema-object-rules) (https://github.com/eslint/rfcs/pull/85) for these two scenarios:

- Drop support for deprecated function-style rules in favor of object style rules
- Require rules with options to specify a schema

This is targeted to be merged as a breaking change in ESLint v9. It's a follow-up to @snitin315's work to add warnings in #16063.

Notable changes:

- Rule tester now throws instead of warns for the two scenarios
- Core no longer normalizes function-style rules into object-style rules
- Core throws when attempting to create a function-style rule (and added a test case for this)
- Retrieving a schema for a rule now defaults to the “no options” schema when no schema is specified
- Updates documentation to mention the new requirements
- I extracted HUNDREDS of updates to test cases that were using function-style rules in #16618
- Throw with no-op schema `schema: {}`
- Implement, document, and test `schema: false` opt-out feature
- Ensure all rule examples in docs that have options also have a schema

#### Is there anything you'd like reviewers to focus on?

Should we delete `docs/developer-guide/working-with-rules-deprecated.md`. Perhaps we should delete most of the content and keep a barebones doc in place to avoid breaking links.

I'm a bit unsure about the below [comment](https://github.com/eslint/eslint/pull/16063#issuecomment-1193192107) from an earlier PR. Should I duplicate the tests into FlatRuleTester? Should we still be relying on asserts in rule tester at all?

> > Should we also duplicate these changes in FlatRuleTester?
>
> I think it isn't necessary since we are planning https://github.com/eslint/rfcs/pull/85 for ESLint v9, and FlatRuleTester will not replace RuleTester before ESLint v9. When we implement https://github.com/eslint/rfcs/pull/85 in linter/config, these deprecation warnings will become errors thrown by linter/config.

### CODE REVIEW

**Comment by nzakas** [comment]
I think it may be a bit too early to be filing a breaking PR when we haven't even planned ESLint v9 yet.

Switching to a draft to avoid accidental merging, but I think it may be better to close this and hold on to the changes until we are actively planning for a major release.

File: tests/lib/rule-tester/rule-tester.js [hunk 1]
```diff
@@ -2376,24 +2335,37 @@ describe("RuleTester", () => {
...
+                invalid: [
+                    { code: "var foo = bar;", options: [{ foo: true }], errors: 1 }
                 ]
-            );
+            }),
+            "`schema: {}` is a no-op. For rules with options, please fill in a complete schema. For rules without options, please omit `schema` or use `schema: []`.");
         });
 
-        it("should not log a deprecation warning when schema is an empty array", () => {
+        it("should not throw an error when schema is an empty array", () => {
```
Comment (ddzz) [inline tests/lib/rule-tester/rule-tester.js:2368]: Seems like these tests could be updated to remove the use of `should`.

Comment (bmish) [reply tests/lib/rule-tester/rule-tester.js:2368]: How else would you write it? I do want it to read like an English sentence.

**Comment by bmish** [comment]
@nzakas "draft" status works for me as I'd love to begin receiving and responding to feedback on this. I will likely have 4 breaking change RFCs to implement for ESLint v9, so I wanted to start preparing them now so they are lined up and ready to go whenever you decide to schedule the release. My concern is that each of these PRs could take several weeks of work, and if I wait until ESLint v9 is announced, I won't have time to complete them.

**Comment by Rec0iL99** [comment]
@bmish wants this open for feedback. Don't close.

**Comment by nzakas** [comment]
@bmish this PR seems very out of date and hasn't been updated in a while. What do you think about closing this for now? You can always pick up the work and submit a new PR later.

**Comment by nzakas** [comment]
Okay, we are just going to close this. @bmish you can feel free to resubmit when it's time.

**Pull request closed without merging**